// bits-ui/dist/bits/toggle-group/toggle-group.svelte.js
import { useRefById } from "svelte-toolbelt";
import { getAriaChecked, getAriaPressed, getDataDisabled, getDataOrientation, getDisabled, } from "../../internal/attrs.js";
import { kbd } from "../../internal/kbd.js";
import { useRovingFocus, } from "../../internal/use-roving-focus.svelte.js";
import { createContext } from "../../internal/create-context.js";
const ROOT_ATTR = "data-toggle-group-root";
const ITEM_ATTR = "data-toggle-group-item";
class ToggleGroupBaseState {
    id;
    ref;
    disabled;
    rovingFocus;
    loop;
    orientation;
    rovingFocusGroup;
    constructor(props) {
        this.id = props.id;
        this.ref = props.ref;
        this.disabled = props.disabled;
        this.rovingFocus = props.rovingFocus;
        this.loop = props.loop;
        this.orientation = props.orientation;
        this.rovingFocusGroup = useRovingFocus({
            candidateAttr: ITEM_ATTR,
            rootNodeId: this.id,
            loop: this.loop,
            orientation: this.orientation,
        });
        useRefById({
            id: this.id,
            ref: this.ref,
        });
    }
    props = $derived.by(() => ({
        id: this.id.current,
        [ROOT_ATTR]: "",
        role: "group",
        "data-orientation": getDataOrientation(this.orientation.current),
        "data-disabled": getDataDisabled(this.disabled.current),
    }));
}
class ToggleGroupSingleState extends ToggleGroupBaseState {
    #value;
    isMulti = false;
    anyPressed = $derived.by(() => this.#value.current !== "");
    constructor(props) {
        super(props);
        this.#value = props.value;
    }
    includesItem(item) {
        return this.#value.current === item;
    }
    toggleItem(item, id) {
        if (this.includesItem(item)) {
            this.#value.current = "";
        }
        else {
            this.#value.current = item;
            this.rovingFocusGroup.setCurrentTabStopId(id);
        }
    }
}
class ToggleGroupMultipleState extends ToggleGroupBaseState {
    #value;
    isMulti = true;
    anyPressed = $derived.by(() => this.#value.current.length > 0);
    constructor(props) {
        super(props);
        this.#value = props.value;
    }
    includesItem(item) {
        return this.#value.current.includes(item);
    }
    toggleItem(item, id) {
        if (this.includesItem(item)) {
            this.#value.current = this.#value.current.filter((v) => v !== item);
        }
        else {
            this.#value.current = [...this.#value.current, item];
            this.rovingFocusGroup.setCurrentTabStopId(id);
        }
    }
}
class ToggleGroupItemState {
    #id;
    #ref;
    #root;
    #value;
    #disabled;
    #isDisabled = $derived.by(() => this.#disabled.current || this.#root.disabled.current);
    constructor(props) {
        this.#value = props.value;
        this.#disabled = props.disabled;
        this.#root = props.rootState;
        this.#id = props.id;
        this.#ref = props.ref;
        useRefById({
            id: this.#id,
            ref: this.#ref,
        });
        $effect(() => {
            if (!this.#root.rovingFocus.current) {
                this.#tabIndex = 0;
            }
            else {
                this.#tabIndex = this.#root.rovingFocusGroup.getTabIndex(this.#ref.current);
            }
        });
        this.onclick = this.onclick.bind(this);
        this.onkeydown = this.onkeydown.bind(this);
    }
    #toggleItem() {
        if (this.#isDisabled)
            return;
        this.#root.toggleItem(this.#value.current, this.#id.current);
    }
    onclick(_) {
        if (this.#isDisabled)
            return;
        this.#root.toggleItem(this.#value.current, this.#id.current);
    }
    onkeydown(e) {
        if (this.#isDisabled)
            return;
        if (e.key === kbd.ENTER || e.key === kbd.SPACE) {
            e.preventDefault();
            this.#toggleItem();
            return;
        }
        if (!this.#root.rovingFocus.current)
            return;
        this.#root.rovingFocusGroup.handleKeydown(this.#ref.current, e);
    }
    isPressed = $derived.by(() => this.#root.includesItem(this.#value.current));
    #ariaChecked = $derived.by(() => {
        return this.#root.isMulti ? undefined : getAriaChecked(this.isPressed, false);
    });
    #ariaPressed = $derived.by(() => {
        return this.#root.isMulti ? getAriaPressed(this.isPressed) : undefined;
    });
    #tabIndex = $state(0);
    props = $derived.by(() => ({
        id: this.#id.current,
        role: this.#root.isMulti ? undefined : "radio",
        tabindex: this.#tabIndex,
        "data-orientation": getDataOrientation(this.#root.orientation.current),
        "data-disabled": getDataDisabled(this.#isDisabled),
        "data-state": getToggleItemDataState(this.isPressed),
        "data-value": this.#value.current,
        "aria-pressed": this.#ariaPressed,
        "aria-checked": this.#ariaChecked,
        disabled: getDisabled(this.#isDisabled),
        [ITEM_ATTR]: "",
        //
        onclick: this.onclick,
        onkeydown: this.onkeydown,
    }));
}
//
// HELPERS
//
function getToggleItemDataState(condition) {
    return condition ? "on" : "off";
}
//
// CONTEXT METHODS
//
const [setToggleGroupRootContext, getToggleGroupRootContext] = createContext("ToggleGroup.Root");
export function useToggleGroupRoot(props) {
    const { type, ...rest } = props;
    const rootState = type === "single"
        ? new ToggleGroupSingleState(rest)
        : new ToggleGroupMultipleState(rest);
    return setToggleGroupRootContext(rootState);
}
export function useToggleGroupItem(props) {
    const rootState = getToggleGroupRootContext();
    return new ToggleGroupItemState({ ...props, rootState });
}
